gcc/jit/jit-playback.c

```diff
@@ -496,6 +497,30 @@ new_function (location *loc,
   /* FIXME: this uses input_location: */
   tree fndecl = build_fn_decl (name, fn_type);
 
+  if (inline_mode != GCC_JIT_INLINE_MODE_DEFAULT)
+  {
+    switch (inline_mode)
+    {
+        case GCC_JIT_INLINE_MODE_DEFAULT:
+            break;
+        case GCC_JIT_INLINE_MODE_ALWAYS_INLINE:
+            DECL_DECLARED_INLINE_P (fndecl) = 1;
+            DECL_DISREGARD_INLINE_LIMITS (fndecl) = 1; // FIXME: this line breaks compilation of sysroot, but seems required to make inlining work.
```

Comment:
There is a recursive `#[inline(always)]` function in libcore. Maybe that is the problem? In rust even `#[inline(always)]` is just a hint that should be ignored if it is impossible to inline.

Reply:
As a quick workaround, we could check in `rustc_cg_gcc` if the function is recursive and then refuse to emit any inline attributes for it. I highly doubt that `DECL_DISREGARD_INLINE_LIMITS` is a good idea here. I'll test tomorrow.

Reply:
@Commeownist Why not? The doc says `This is equivalent to using the always_inline attribute without the required diagnostics if the function cannot be inlined.`, but I'm not sure what it actually means. I'll have to read more code.

Reply:
@antoyo `DECL_DISREGARD_INLINE_LIMITS` is totally correct here, see [this function](https://github.com/gcc-mirror/gcc/blob/1932e1169a236849f5e7f1cd386da100d9af470f/gcc/ipa-inline.c#L2834). My mistake, confused it with something else.

Reply:
This function seems to actually check whether the function being inlined is [recursive](https://github.com/gcc-mirror/gcc/blob/1932e1169a236849f5e7f1cd386da100d9af470f/gcc/ipa-inline.c#L2845).